File: 141_linked_list_cycle/solution.py
# Definition for singly-linked list.
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

    def __str__(self):
        return str(self.value)

    def has_cycle(self):
        visited = set()
        curr = self
        while curr:
            if curr.next is None:
                return False
            if curr in visited:
                return True
            visited.add(curr)
            curr = curr.next


mon = Node("monday")
tue = Node("tuesday")
wed = Node("wednesday")
thu = Node("thursday")
fri = Node("friday")
sat = Node("saturday")
sun = Node("sunday")

mon.next = tue
tue.next = wed
wed.next = thu
thu.next = fri
fri.next = sat
sat.next = sun
sun.next = mon

print(mon.has_cycle())

File: 141_linked_list_cycle/solution.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func hasCycle(head *ListNode) bool {
	var fast = head
	var slow = head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			return true
		}
	}
	return false
}

func main() {
	mon := ListNode{Val: 1}
	tue := ListNode{Val: 2}
	wed := ListNode{Val: 3}
	thu := ListNode{Val: 4}
	fri := ListNode{Val: 5}
	sat := ListNode{Val: 6}
	sun := ListNode{Val: 7}

	mon.Next = &tue
	tue.Next = &wed
	wed.Next = &thu
	thu.Next = &fri
	fri.Next = &sat
	sat.Next = &sun
	sun.Next = &mon

	fmt.Println(hasCycle(&mon))
}
